subcommand/server-acl-init/command.go

```diff
@@ -244,13 +244,48 @@ func (c *Command) getConsulServers(logger hclog.Logger) *apiv1.PodList {
 			}
 			return nil
 		}, logger)
-	return serverPods
+
+	var podAddrs []podAddr
+	for _, pod := range serverPods.Items {
+		var httpPort int32
+		for _, p := range pod.Spec.Containers[0].Ports {
+			if p.Name == "http" {
+				httpPort = p.ContainerPort
+			}
+		}
+		if httpPort == 0 {
+			return nil, fmt.Errorf("pod %s has no port labeled 'http'", pod.Name)
+		}
+		addr := fmt.Sprintf("%s:%d", pod.Status.PodIP, httpPort)
+		podAddrs = append(podAddrs, podAddr{
+			Name: pod.Name,
+			Addr: addr,
+		})
+	}
+	return podAddrs, nil
 }
 
-func (c *Command) bootstrapACLs(logger hclog.Logger, consulClient *api.Client) ([]byte, error) {
-	// Bootstrap the ACLs unless already bootstrapped.
-	alreadyBootstrapped := false
+// bootstrapServers bootstraps ACLs and ensures each server has an ACL token.
+func (c *Command) bootstrapServers(logger hclog.Logger, bootTokenSecretName string) (string, error) {
+	serverPods, err := c.getConsulServers(logger, c.flagReplicas)
+	if err != nil {
+		return "", err
+	}
+	logger.Info(fmt.Sprintf("Found %d Consul server Pods", len(serverPods)))
+
+	// Pick the first pod to connect to for bootstrapping and set up connection.
+	firstServerAddr := serverPods[0].Addr
+	consulClient, err := api.NewClient(&api.Config{
+		Address: firstServerAddr,
+		Scheme:  "http",
+	})
+	if err != nil {
+		return "", fmt.Errorf("creating Consul client for address %s: %s", firstServerAddr, err)
+	}
+
+	// Call bootstrap ACLs API.
```

Comment:
I think we need to move this into the `c.untilSucceeds` to avoid a race condition when the ACL init job discovers the server that is about to be terminated.

Reply:
Hmm, that case should now not be possible but in another way it could happen. Previously, if this job was run again (ex. during a helm upgrade), it would try to bootstrap servers again, even if there was the Kubernetes bootstrap secret.

Now, during a helm upgrade it would not try to boostrap again because it will detect the secret.

However, say they added the meshgateway=true value and ran an upgrade. Then [here](https://github.com/hashicorp/consul-k8s/blob/5a1b0928f789e50b99d6c9492e6e0c6bf7c9d8e4/subcommand/server-acl-init/command.go#L141) we grab the first consul server and use it to generate tokens. If that server was lost then we'd get into the same situation.

So maybe we need to loop back up and try a different server in that situation. I'll look into how best to do that in the code.

Reply:
Yep, you're right! I keep coming back to this. Although I think this could happen during initial installation too. For example, suppose your servers didn't come up successfully and you had to do something to restart them. It would be nice if the acl init job (which would be already running) could recover on its own. 

Your example with enabling a feature that requires an ACL change is a great example!

Reply:
I've added code that checks if there's an in-progress rollout and waits for it to finish before proceeding. I've also added an overall timeout so if we get into a pathological situation where we're looping forever we'll eventually exit and the job will retry.

Reply:
FYI if you're testing this use my latest PR of consul-helm because the acl-init job needs rbac for listing the statefulsets

Reply:
> It would be nice if the acl init job (which would be already running) could recover on its own.

I forgot to address this point. I agree that this would be nice but I think there's too much complexity to make it happen. We'd need to be able to re-lookup consul server pod ips and recreate a consul client given a certain error condition. Since the API calls happen so fast once we get a consul IP that works I think it's sufficient to wait for a rollout to be complete and then continue.

However I do agree that there are cases that we might get into and rather than add a layer of complexity for relooping back to create the consul client I added an overall timeout mechanism for the command which should then get retried because its a job.

Not saying that's the only way of doing things though.